// angular-svg-round-progressbar/esm2022/round-progress/round-progress.module.mjs
import { NgModule } from '@angular/core';
import { RoundProgressComponent } from './round-progress.component';
import { ROUND_PROGRESS_DEFAULTS_PROVIDER } from './round-progress.config';
import * as i0 from "@angular/core";
class RoundProgressModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.4", ngImport: i0, type: RoundProgressModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.4", ngImport: i0, type: RoundProgressModule, declarations: [RoundProgressComponent], exports: [RoundProgressComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.4", ngImport: i0, type: RoundProgressModule, providers: [ROUND_PROGRESS_DEFAULTS_PROVIDER] }); }
}
export { RoundProgressModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.4", ngImport: i0, type: RoundProgressModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RoundProgressComponent],
                    exports: [RoundProgressComponent],
                    providers: [ROUND_PROGRESS_DEFAULTS_PROVIDER],
                }]
        }] });
